api/valor_api/backend/metrics/metric_utils.py

```diff
@@ -253,6 +256,109 @@ def create_metric_mappings(
     return ret
 
 
+def log_evaluation_analytics(
+    db: Session,
+    evaluation_id: int,
+    prediction_filter: schemas.Filter,
+    groundtruth_filter: schemas.Filter,
+):
+    """
+    Store analytics regarding the evaluation's execution in the metadata field of the evaluation table.
+
+    Parameters
+    ----------
+    db : Session
+        The database Session to query against.
+    evaluation_id : int
+        The job ID to create metrics for.
+    prediction_filter : schemas.Filter
+        The filter to be used to query predictions.
+    groundtruth_filter : schemas.Filter
+        The filter to be used to query groundtruths.
+    """
+    # get ground truth, prediction, annotation, and label counts
+    gt_subquery = (
+        Query(
+            models.GroundTruth,
+        )
+        .filter(groundtruth_filter)
+        .groundtruths(as_subquery=False)
+        .alias()
+    )
+
+    gts = db.execute(
+        select(
+            gt_subquery.c.annotation_id,
+            gt_subquery.c.label_id,
+        ).select_from(gt_subquery)
+    ).all()
+
+    # handle edge case where no gts come back
+    if not gts:
+        gt_annotation_id, gt_label_id = set(), set()
+    else:
+        gt_annotation_id, gt_label_id = map(set, zip(*gts))
+
+    pd_subquery = (
+        Query(
+            models.Prediction,
+        )
+        .filter(prediction_filter)
+        .predictions(as_subquery=False)
+        .alias()
+    )
+
+    pds = db.execute(
+        select(
+            pd_subquery.c.annotation_id,
+            pd_subquery.c.label_id,
+        ).select_from(pd_subquery)
+    ).all()
+
+    if not pds:
+        pd_annotation_id, pd_label_id = set(), set()
+    else:
+        pd_annotation_id, pd_label_id = map(set, zip(*pds))
+
+    annotation_cnt = len(gt_annotation_id | pd_annotation_id)
+    label_cnt = len(gt_label_id | pd_label_id)
+
+    # get datum count
+    pd_datums = db.query(
+        Query(models.Dataset.name, models.Datum.id)  # type: ignore - sqlalchemy issues
+        .filter(prediction_filter)
+        .predictions()
+    ).all()
+    gt_datums = db.query(
+        Query(models.Dataset.name, models.Datum.id)  # type: ignore - sqlalchemy issues
+        .filter(groundtruth_filter)
+        .groundtruths()
+    ).all()
+    unique_datums = set(pd_datums + gt_datums)
+    datum_cnt = len(unique_datums)
+
+    evaluation = core.fetch_evaluation_from_id(db, evaluation_id)
+
+    server_time = db.execute(func.now()).scalar().replace(tzinfo=None)  # type: ignore - guaranteed to return server time if psql is running
```

Comment:
Why do we need this if an evaluation already has a `created_at` timestamp?

Reply:
to calculate duration, we need to compare the server time at completion to the `created_at` time